### PR TITLE
feat: [PL-41637]: Delegate notification template registration

### PR DESCRIPTION
**Pull request by jennyjames21**
## Describe your changes
Summary:
- Notification templates for delegate notification events(delegate down, delegate expired, delegate about to expire) for all five channels(email, slack, msteams, pagerduty, webhook)

Test
- Verified  templates got registered with notification service
image

- Template format for email
image

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/55194)

### CODE REVIEW

**Comment by jennyjames21** [comment]
trigger uts

**Comment by jennyjames21** [comment]
trigger uts

File: 400-rest/src/main/java/io/harness/notifications/NotificationTemplateRegistrar.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        PredefinedTemplate.DELEGATE_DOWN_SLACK, PredefinedTemplate.DELEGATE_EXPIRED_SLACK,
+        PredefinedTemplate.DELEGATE_ABOUT_EXPIRE_SLACK, PredefinedTemplate.DELEGATE_DOWN_MSTEAMS,
+        PredefinedTemplate.DELEGATE_EXPIRED_MSTEAMS, PredefinedTemplate.DELEGATE_ABOUT_EXPIRE_MSTEAMS,
+        PredefinedTemplate.DELEGATE_DOWN_PAGERDUTY, PredefinedTemplate.DELEGATE_EXPIRED_PAGERDUTY,
+        PredefinedTemplate.DELEGATE_ABOUT_EXPIRE_PAGERDUTY, PredefinedTemplate.DELEGATE_DOWN_WEBHOOK,
+        PredefinedTemplate.DELEGATE_EXPIRED_WEBHOOK, PredefinedTemplate.DELEGATE_ABOUT_EXPIRE_WEBHOOK);
+    for (PredefinedTemplate template : templates) {
+      try {
+        log.info("Registering {} with NotificationService", template.getIdentifier());
+        notificationClient.saveNotificationTemplate(Team.OTHER, template, true);
```
Comment (xingchi-jin) [inline 400-rest/src/main/java/io/harness/notifications/NotificationTemplateRegistrar.java:38]: lack some background knowledge, where is the template saved to ? and where the template is used ?

**Review by xingchi-jin — DISMISSED** [review]

**Comment by xingchi-jin** [comment]
btw, in the scrrnshot, the delegate disconnect email notification looks like a celebration  : P

File: 400-rest/src/main/java/io/harness/notifications/NotificationTemplateRegistrar.java (hunk 1 above)
Comment (jennyjames21) [reply 400-rest/src/main/java/io/harness/notifications/NotificationTemplateRegistrar.java:38]: Templates saved in notification service DB on template registration
When the client sends the notification trigger to the notification service,  service will fetch the customer rule, and based on the channel type(email, slack) it fetches the template from DB to send the notification to customer.

File: 400-rest/src/main/resources/notificationtemplates/webhook/delegate_about_to_expire.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{
+    "text": "Harness Delegate ${DELEGATE_ID} expires on ${EXPIRY_TIME}. Make sure auto upgrade is ON to get the latest image."
```
Comment (raghuAtWings) [inline 400-rest/src/main/resources/notificationtemplates/webhook/delegate_about_to_expire.txt:2]: we should change the text to

Harness Delegate ${DELEGATE_ID} expires on ${EXPIRY_TIME}. Either set auto upgrade is ON or upgrade to latest version. More information at https://developer.harness.io/docs/platform/delegates/install-delegates/delegate-upgrades-and-expiration/

File: 400-rest/src/main/resources/notificationtemplates/webhook/delegate_expired.txt [hunk 3]
```diff
@@ -0,0 +1,3 @@
+{
+    "text": "Harness Delegate ${DELEGATE_ID} expired on ${EXPIRY_TIME}. Upgrade the delegate"
```
Comment (raghuAtWings) [inline 400-rest/src/main/resources/notificationtemplates/webhook/delegate_expired.txt:2]: Harness Delegate ${DELEGATE_ID} expired on ${EXPIRY_TIME}. Upgrade the delegate. More infromation at https://developer.harness.io/docs/platform/delegates/install-delegates/delegate-upgrades-and-expiration/

File: 400-rest/src/main/resources/notificationtemplates/slack/delegate_expired.txt [hunk 4]
```diff
@@ -0,0 +1,3 @@
+{
+    "text": "Harness Delegate ${DELEGATE_ID} expired on ${EXPIRY_TIME}. Upgrade the delegate"
```
Comment (raghuAtWings) [inline 400-rest/src/main/resources/notificationtemplates/slack/delegate_expired.txt:2]: are we sending delegate id or name?

**Review by raghuAtWings — CHANGES_REQUESTED** [review]

File: 400-rest/src/main/resources/notificationtemplates/slack/delegate_expired.txt (hunk 4 above)
Comment (jennyjames21) [reply 400-rest/src/main/resources/notificationtemplates/slack/delegate_expired.txt]: updated

File: 400-rest/src/main/resources/notificationtemplates/webhook/delegate_expired.txt (hunk 3 above)
Comment (jennyjames21) [reply 400-rest/src/main/resources/notificationtemplates/webhook/delegate_expired.txt]: updated

**Comment by jennyjames21** [comment]
trigger messagecheck

**Review by raghuAtWings — DISMISSED** [review]

**Comment by jennyjames21** [comment]
trigger uts

**Comment by jennyjames21** [comment]
trigger uts

**Comment by jennyjames21** [comment]
trigger uts

File: platform-service/modules/notification-service/src/main/resources/templates/email/delegate_about_to_expire.txt [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+                  <img border="0" vspace="0" hspace="0" src="https://s3.amazonaws.com/wings-assets/ng-mail-assets/ng-background.png" alt="Please enable images to view this content" title="Harness Hero" width="580" style="line-height: 100%; width: 100%; max-width: 580px; color: #000000; font-size: 13px; margin: 0; padding: 0; outline: none; text-decoration: none; -ms-interpolation-mode: bicubic; border: none; display: block;">
+                </a>
+              </td>
+            </tr>
+            <tr>
+              <td align="center" valign="middle" style="-webkit-font-smoothing: antialiased; text-size-adjust: 100%; -ms-text-size-adjust: 100%; -webkit-text-size-adjust: 100%; line-height: 100%; mso-table-lspace: 0pt; mso-table-rspace: 0pt; border-spacing: 0; margin: 0; padding: 0; padding-left: 6.25%; padding-right: 6.25%; width: 87.5%; padding-top: 0px; height: 70px; border-collapse: collapse;" height="70" class="button" width="87.5%">
+                Delegate ${delegateId} is getting expired on ${expiryDate}. Please make sure auto grade is ON to get the latest build image.
```
Comment (vanshBhatia-A4k9) [inline platform-service/modules/notification-service/src/main/resources/templates/email/delegate_about_to_expire.txt:41]:
```suggestion
                Delegate ${delegateId} is getting expired on ${expiryDate}. Please make sure auto upgrade is ON to get the latest build image.
```

**Comment by jennyjames21** [comment]
trigger uts

**Comment by jennyjames21** [comment]
trigger ut4

**Review by raghuAtWings — APPROVED** [review]

**Pull request merged**